src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,199 @@
+
     @Override
     public void add(T value) {
+        linkLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        areIndexInSizeRange(index);
+        linkReassignmentForAdd(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        final T removedItem = node(index).item;
+        node(index).item = value;
+        return removedItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        final T removedItem = node(index).item;
+        linkReassignmentForRemove(index);
+        return removedItem;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = findIndex(object);
+        if (index == FALSE_INDEX) {
+            return false;
+        }
+        linkReassignmentForRemove(index);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void areIndexInSizeRange(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + " doesn't exist");
+        }
+    }
+
+    private void areIndexInRange(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + " doesn't exist");
+        }
+    }
+
+    private int findIndex(T object) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            T currentObject = node.item;
+            if (areItemsEqual(currentObject, object)) {
+                return i;
+            } else {
+                node = node.next;
+            }
+        }
+        return FALSE_INDEX;
+    }
+
+    Node<T> node(int index) {
+        areIndexInRange(index);
+        Node<T> wantedNode;
+        if (index < (size >> 1)) {
+            wantedNode = first;
+            for (int i = 0; i < index; i++) {
+                wantedNode = wantedNode.next;
+            }
+        } else {
+            wantedNode = last;
+            for (int i = size - 1; i > index; i--) {
+                wantedNode = wantedNode.prev;
+            }
+        }
+        return wantedNode;
+    }
+
+    private void linkReassignmentForAdd(T value, int index) {
+        if (index == size) {
+            linkLast(value);
+            return;
+        }
+        Node<T> indexNode = node(index);
+        if (indexNode.prev == null) {
+            linkFirst(value, indexNode);
+        } else {
+            linkBefore(value, indexNode);
+        }
+    }
+
+    private void linkReassignmentForRemove(int index) {
```

Comment:
rename method as `unlink()` that was mentioned in common mistakes, and put as parameter `Node<T>` instead of `index`
  `private T unlink(Node<T> node) `
refactor the method body without usage of `index`.  
refactor methods  `remove(T object)` and `remove(int index)` that called `unlink(node)` accordingly